### PR TITLE
skip to replace link text which is a link

### PR DESCRIPTION
**Pull request by eh2077**
### Fixed Issues

$ https://github.com/Expensify/App/issues/16475
PROPOSAL: https://github.com/Expensify/App/issues/16475#issuecomment-1492798632

# Tests
1. Go to a chat and add a new comment, like `test`
2. Edit the comment to replace it with the following link and save it
```
https://www.facebook.com/hashtag/__main/?__eep__=6
```
3. Verify the link text and url are same as the original link
4. Edit the link to remove the link by replacing the link markdown the following original link and save it
```
https://www.facebook.com/hashtag/__main/?__eep__=6
```
5. Verify the link is not autolinked and displayed as text `https://www.facebook.com/hashtag/__main/?__eep__=6`

- [ ] Verify that no errors appear in the JS console

### QA Steps

Same as test.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>

<summary>Web</summary>

https://user-images.githubusercontent.com/117511920/232646737-a850e398-97da-4294-b5d3-a66715b87fe6.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/117511920/232646851-dc22a170-0699-477b-b0e1-dfc49cb1da7a.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/117511920/232646930-0afddf6a-7561-48aa-ba2b-e0a5ca53b5ae.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/117511920/232646960-5b0950cf-fd2d-4c08-8e30-bfe92e0e96eb.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/117511920/232647019-9e8ba7e1-8c2f-46d7-8581-471b62a3aa13.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/117511920/232647035-cd7fc875-d703-4ed4-bd53-b91dc1cdaff3.mp4

</details>

### CODE REVIEW

File: tests/actions/ReportTest.js [hunk 1]
```diff
@@ -430,40 +430,57 @@ describe('actions/Report', () => {
...
+        newCommentMarkdown = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
         expectedOutput = 'Comment www.google.com';
         expect(newCommentMarkdown).toBe(expectedOutput);
 
         // User Delete only () part of link but leaves the []
         // We should not generate link
         originalCommentHTML = 'Comment <a href="https://www.google.com" target="_blank">www.google.com</a>';
         afterEditCommentText = 'Comment [www.google.com]';
-        newCommentMarkdown = Report.handleUserDeletedLinks(afterEditCommentText, originalCommentHTML);
+        newCommentMarkdown = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
```
Comment (tienifr) [inline tests/actions/ReportTest.js:449]: Should we change the name `newCommentMarkdown` here to `newCommentHtml`

File: tests/actions/ReportTest.js [hunk 2]
```diff
@@ -430,40 +430,57 @@ describe('actions/Report', () => {
...
         expectedOutput = 'Comment [www.google.com]';
         expect(newCommentMarkdown).toBe(expectedOutput);
 
         // User Generates multiple links in one edit
         // We should generate both links
         originalCommentHTML = 'Comment';
         afterEditCommentText = 'Comment www.google.com www.facebook.com';
-        newCommentMarkdown = Report.handleUserDeletedLinks(afterEditCommentText, originalCommentHTML);
-        expectedOutput = 'Comment [www.google.com](https://www.google.com) [www.facebook.com](https://www.facebook.com)';
+        newCommentMarkdown = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
```
Comment (tienifr) [inline tests/actions/ReportTest.js:457]: same as above

File: tests/actions/ReportTest.js [hunk 3]
```diff
@@ -430,40 +430,57 @@ describe('actions/Report', () => {
...
+            + '<a href="https://www.facebook.com" target="_blank" rel="noreferrer noopener">www.facebook.com</a>';
         expect(newCommentMarkdown).toBe(expectedOutput);
 
         // Comment has two links but user deletes only one of them
         // Should not generate link again for the deleted one
         originalCommentHTML = 'Comment <a href="https://www.google.com" target="_blank">www.google.com</a>  <a href="https://www.facebook.com" target="_blank">www.facebook.com</a>';
         afterEditCommentText = 'Comment www.google.com  [www.facebook.com](https://www.facebook.com)';
-        newCommentMarkdown = Report.handleUserDeletedLinks(afterEditCommentText, originalCommentHTML);
-        expectedOutput = 'Comment www.google.com  [www.facebook.com](https://www.facebook.com)';
+        newCommentMarkdown = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
```
Comment (tienifr) [inline tests/actions/ReportTest.js:466]: same as above

File: tests/actions/ReportTest.js [hunk 4]
```diff
@@ -430,40 +430,57 @@ describe('actions/Report', () => {
...
+        expectedOutput = 'Comment www.google.com  <a href="https://www.facebook.com" target="_blank" rel="noreferrer noopener">www.facebook.com</a>';
+        expect(newCommentMarkdown).toBe(expectedOutput);
+
+        // User edit and replace comment with link include underscore
+        // We should generate link and the link text is same as the link
+        originalCommentHTML = 'Comment';
+        afterEditCommentText = 'https://www.facebook.com/hashtag/__main/?__eep__=6';
+        newCommentMarkdown = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
```
Comment (tienifr) [inline tests/actions/ReportTest.js:474]: same as above

File: tests/actions/ReportTest.js (hunk 1 above)
Comment (eh2077) [reply tests/actions/ReportTest.js]: updated

File: package.json [hunk 5]
```diff
@@ -69,7 +69,7 @@
     "babel-polyfill": "^6.26.0",
     "dom-serializer": "^0.2.2",
     "domhandler": "^4.3.0",
-    "expensify-common": "git+ssh://git@github.com/Expensify/expensify-common.git#be625f55058111793bbcfe3f64788ea0a8f9d705",
+    "expensify-common": "git+https://github.com/eh2077/expensify-common.git#540a9451c0452277151b2a63ad46c9591da4e3f5",
```
[the remark this answers is not in the document]
Comment (eh2077) [reply package.json]: Updated

**Review by fedirjh — DISMISSED** [review]

File: src/libs/actions/Report.js [hunk 6]
```diff
@@ -823,42 +823,42 @@ const getRemovedMarkdownLinks = (oldComment, newComment) => {
...
-const removeLinks = (comment, links) => {
-    let commentCopy = comment.slice();
+const removeLinksFromHtml = (html, links) => {
+    let htmlCopy = html.slice();
     _.forEach(links, (link) => {
-        const regex = new RegExp(`\\[([^\\[\\]]*)\\]\\(${link}\\)`, 'gm');
-        const linkMatch = regex.exec(commentCopy);
-        const linkText = linkMatch && linkMatch[1];
-        commentCopy = commentCopy.replace(`[${linkText}](${link})`, linkText);
+        const regex = new RegExp(`<(a)[^><]*href\\s*=\\s*(['"])(${Str.escapeForRegExp(link)})\\2(?:".*?"|'.*?'|[^'"><])*>([\\s\\S]*?)<\\/\\1>(?![^<]*(<\\/pre>|<\\/code>))`, 'gi');
```
[the remark this answers is not in the document]
Comment (eh2077) [reply src/libs/actions/Report.js:841]: @fedirjh Updated

File: tests/actions/ReportTest.js [hunk 7]
```diff
@@ -430,41 +430,74 @@ describe('actions/Report', () => {
...
+
+        // User edit and delete the link with underscore
+        // We should not generate link
+        originalCommentHTML = '<a href="https://www.facebook.com/hashtag/__main/?__eep__=6" target="_blank" rel="noreferrer noopener">https://www.facebook.com/hashtag/__main/?__eep__=6</a>';
+        afterEditCommentText = 'https://www.facebook.com/hashtag/__main/?__eep__=6';
+        newCommentHTML = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
+        expectedOutput = 'https://www.facebook.com/hashtag/__main/?__eep__=6';
+        expect(newCommentHTML).toBe(expectedOutput);
+
+        // User edit and replace comment with link include asterisk
```
Comment (amyevans) [inline tests/actions/ReportTest.js:486]:
```suggestion
        // User edits and replaces comment with a link containing asterisks
```

File: tests/actions/ReportTest.js [hunk 8]
```diff
@@ -430,41 +430,74 @@ describe('actions/Report', () => {
...
+        // We should generate link
+        originalCommentHTML = 'Comment';
+        afterEditCommentText = 'http://example.com/foo/*/bar/*/test.txt';
+        newCommentHTML = Report.handleUserDeletedLinksInHtml(afterEditCommentText, originalCommentHTML);
+        expectedOutput = '<a href="http://example.com/foo/*/bar/*/test.txt" target="_blank" rel="noreferrer noopener">http://example.com/foo/*/bar/*/test.txt</a>';
+        expect(newCommentHTML).toBe(expectedOutput);
+
+        // User edit and delete the link with asterisk
```
Comment (amyevans) [inline tests/actions/ReportTest.js:494]:
```suggestion
        // User edits and deletes the link containing asterisks
```

**Review by amyevans — CHANGES_REQUESTED** [review]
Looks good, just some suggestions for improving comment clarity

File: src/libs/actions/Report.js [hunk 9]
```diff
@@ -2,8 +2,8 @@ import {InteractionManager} from 'react-native';
 import _ from 'underscore';
 import lodashGet from 'lodash/get';
 import ExpensiMark from 'expensify-common/lib/ExpensiMark';
-import Onyx from 'react-native-onyx';
 import Str from 'expensify-common/lib/str';
+import Onyx from 'react-native-onyx';
```
[the remark this answers is not in the document]
Comment (eh2077) [reply src/libs/actions/Report.js]: updated

**Review by amyevans — APPROVED** [review]

**Pull request merged**